common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesBrokerApplicationMetadataCache.java

```diff
@@ -51,6 +51,32 @@ public SharedPreferencesBrokerApplicationMetadataCache(@NonNull final Context co
         super(context, DEFAULT_APP_METADATA_CACHE_NAME, KEY_CACHE_LIST);
     }
 
+    @Override
+    public boolean insert(@NonNull final BrokerApplicationMetadata brokerApplicationMetadata) {
+        // Before we insert a record, if we have a existing entry with the same client id, env, and
+        // processUid, we should remove that entry...
+        disposeOfDuplicateRecords(
+                brokerApplicationMetadata.getClientId(),
+                brokerApplicationMetadata.getEnvironment(),
+                brokerApplicationMetadata.getUid()
+        );
+        return super.insert(brokerApplicationMetadata);
+    }
+
+    private void disposeOfDuplicateRecords(@NonNull final String clientId,
+                                           @NonNull final String environment,
```

Comment:
How does this work for cross cloud apps? Could there be records in other environments? In this case I believe they won't be removed? Do we want to remove them though?

Reply:
Since we don't yet support x-cloud apps and I don't have insight into the future caching changes @iamgusain will be making in broker, that's a design question @iamgusain will need to account for in his implementation

Reply:
Assuming they use the same client registration, [this function](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesBrokerApplicationMetadataCache.java#L74-L77) can be used to determine if an app is foci, even x-cloud

Reply:
Sounds good...resolving..

Reply:
Possibly [this function](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesBrokerApplicationMetadataCache.java#L137) gets updated to accept a wildcard environment... I don't think this code will be super hard to adapt; will defer to @iamgusain to adapt this implementation as required when the time comes